requests/sessions.py

```diff
@@ -633,7 +633,7 @@ def send(self, request, **kwargs):
         kwargs.setdefault('stream', self.stream)
         kwargs.setdefault('verify', self.verify)
         kwargs.setdefault('cert', self.cert)
-        kwargs.setdefault('proxies', self.proxies)
+        kwargs.setdefault('proxies', self.rebuild_proxies(request, self.proxies))
```

Comment:
I'm 99% certain this isn't actually correct. This does work regardless of whether it's used or not. We should do:

```py
if kwargs['proxies']:
    kwargs['proxies'] = self.rebuild_proxies(request, kwargs['proxies'])
```

After this line though

Reply:
What do you mean by "this"?

The problem or the proposed solution?

For the problem, I've added a fell test cases, which when I remove this line change those were not passing.

As for the solution.

Should the `proxies` parameter be overwritten by the environment?

If we `rebuild_proxies` there, it will load the environment variables, instead of using what is being passed on the arguments.

My solution is intentionally considering the arguments as the highest priority, however the environment is loaded when redirection occurs [here](https://github.com/psf/requests/pull/5681/files#diff-05180dda5cdc530f95d400932961d6152bd515a6447af23d22c7f335b390b9d8R572) is the test case that highlights that.

Reply:
@sigmavirus24 

Does my answer help you understand the problem in any way?

I'm facing this problem in production and the tests do show the inconsistent use of proxy environment variables by `requests.session`.

Reply:
I don't recall 😆 but your logic makes sense.